### PR TITLE
[MODEL] Add support for Zamba2 models

### PR DESCRIPTION
**Pull request by yury-tokpanov**
This PR adds support for Zamba2 models (#9382), a series of mamba2-transformer hybrid models with shared attention blocks and LoRAs, applied to shared MLP and attention blocks, depending on the model. 1.2B and 7B models use RoPE for their attention blocks.

This PR is fully compatible with Zamba2 integration in HuggingFace transformers library, which was recently merged into the main branch.

- We are able to reproduce evaluation results using evaluation harness with vllm evaluator.
- We also inspected logits and intermediate layers outputs, comparing them with our reference implementation. We find a good agreement between the two (given numerical precision).
- Chunked prefill appears to be working.
- TP is supported.
- PP is not supported, and we believe it wouldn't make much sense to use our models with PP, since it will remove the memory advantage of shared attention layers.

Unit tests pass now.

We would like to acknowledge authors of [Bamba PR](https://github.com/vllm-project/vllm/pull/10909) and [Mamba2 PR](https://github.com/vllm-project/vllm/pull/9292) (@fabianlim and @tlrmchlsmth respectively) for adding mamba2 support to vLLM and having productive discussions!

cc: @Quentin-Anthony @BerenMillidge @pglorio

### CODE REVIEW

File: requirements-common.txt [hunk 1]
```diff
@@ -6,7 +6,7 @@ requests >= 2.26.0
 tqdm
 blake3
 py-cpuinfo
-transformers >= 4.48.2  # Required for Bamba model and Transformers backend.
+transformers >= 4.49.0  # Required for Bamba and Zamba2 models and Transformers backend.
```
Comment (tlrmchlsmth) [inline requirements-common.txt:9]: Typically we've been labeling this with the most recent model rather than appending to this line:
```suggestion
transformers >= 4.49.0  # Required for Zamba2 models and Transformers backend.
```

File: vllm/model_executor/models/zamba2.py [hunk 2]
```diff
@@ -0,0 +1,1078 @@
...
+                                         bias=False,
+                                         quant_config=quant_config,
+                                         gather_output=True),
+                    ColumnParallelLinear(config.adapter_rank,
+                                         2 * self.intermediate_size,
+                                         bias=False,
+                                         quant_config=quant_config),
+                ])
+            else:
+                gate_up_proj_adapter = nn.Identity()
```
Comment (tlrmchlsmth) [inline vllm/model_executor/models/zamba2.py:329]: What do you think about creating a class to encapsulate these lora adapters?  I see some repeated mechanical-looking code that might make sense to factor out.

Comment (yury-tokpanov) [reply vllm/model_executor/models/zamba2.py:329]: I like this idea, I'll do that.

File: vllm/model_executor/models/zamba2.py [hunk 3]
```diff
@@ -0,0 +1,1078 @@
...
+        super().__init__()
+        self.config = config
+        self.tp_size = get_tensor_model_parallel_world_size()
+        self.layer2block_map = layer2block_map
+        self.hidden_size = config.hidden_size
+        self.intermediate_size = config.intermediate_size
+        self.num_fwd_mem_blocks = len(layer2block_map)
+
+        # Main projection layers with gating
+        self.gate_up_proj = ColumnParallelLinear(
```
[the remark this answers is not in the document]
Comment (yury-tokpanov) [reply vllm/model_executor/models/zamba2.py:296]: gate and projector are already merged in our checkpoint into a single tensor, so wouldn't ColumnParallel be enough?

File: vllm/model_executor/models/zamba2.py [hunk 4]
```diff
@@ -0,0 +1,1078 @@
...
+            bias=False,
+            quant_config=quant_config,
+        )
+        self.o_proj = RowParallelLinear(self.attention_hidden_size,
+                                        config.hidden_size,
+                                        bias=False,
+                                        quant_config=quant_config)
+
+        # Need to define separate Attention objects, because in recent vLLM
+        # KV cache tensors are tied to specific Attention objects.
```
[the remark this answers is not in the document]
Comment (yury-tokpanov) [reply vllm/model_executor/models/zamba2.py]: added more comments

File: vllm/model_executor/models/zamba2.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yury-tokpanov) [reply vllm/model_executor/models/zamba2.py:279]: hmm, I could theoretically split weights when loading them (passing splits as shards). I guess using MergedColumnParallel would allow us to avoid doing all-gather before chunk, and we could use RowParallel for down_proj (instead of ReplicatedLinear). However, we wouldn't be saving much memory here anyway (as there are just a few MLP blocks), and all_reduce in RowParallel is actually slower.

[the remark this answers is not in the document]
Comment (fabianlim) [reply vllm/model_executor/models/zamba2.py:279]: `MergedColumnParallel` will take care of sharding of merged weights. This is especially important if say gate and up are merged, since then if TP=2, then the first device will need the first part of gate and up, and the second part willl require the 2nd part.

[the remark this answers is not in the document]
Comment (yury-tokpanov) [reply vllm/model_executor/models/zamba2.py:279]: ok, let me try that. To make things work with TP > 1, I put all_gather after our LoRA (I also had to use ReplicatedLInear for down_proj). I think, it has minimal effect (because we don't many MLP blocks) on memory.

I'm doing profiling now as well to see, where we can improve efficiency.

[the remark this answers is not in the document]
Comment (yury-tokpanov) [reply vllm/model_executor/models/zamba2.py:279]: Thanks for the clarification. Yes, I was confused.

[the remark this answers is not in the document]
Comment (yury-tokpanov) [reply vllm/model_executor/models/zamba2.py:279]: Something is off. When I change to MergedColumnParallelLinear (and then RowColumnParallel for down_proj), eval scores drop.

[the remark this answers is not in the document]
Comment (yury-tokpanov) [reply vllm/model_executor/models/zamba2.py:279]: ok, seems like I figured it out. Yes, needed to update LoRA as well. Going to do more tests and then push new changes.

File: vllm/model_executor/models/zamba2.py (hunk 2 above)
Comment (yury-tokpanov) [reply vllm/model_executor/models/zamba2.py]: done with that

File: vllm/model_executor/models/zamba2.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yury-tokpanov) [reply vllm/model_executor/models/zamba2.py]: done with that

**Comment by yury-tokpanov** [comment]
seems like current failures is due to cv2 imports in transformers v4.49.0. This is a known issue: https://github.com/vllm-project/vllm/pull/13905

**Comment by yury-tokpanov** [comment]
@tlrmchlsmth other than the external issue with the latest released transformers (cv2 import in 4.49.0, but I see it's fixed in their dev branch), do you have other suggestions for this PR?

**Review by tlrmchlsmth — COMMENTED** [review]
The PR looks great to me, thanks for the contribution -- I'll accept once the transformers 4.49 issue is resolved

File: tests/models/registry.py [hunk 5]
```diff
@@ -193,6 +193,7 @@ def check_available_online(
     "XverseForCausalLM": _HfExamplesInfo("xverse/XVERSE-7B-Chat",
                                          is_available_online=False,
                                          trust_remote_code=True),
+    "Zamba2ForCausalLM": _HfExamplesInfo("Zyphra/Zamba2-7B-instruct"),
```
Comment (DarkLight1337) [inline tests/models/registry.py:196]:
```suggestion
    "Zamba2ForCausalLM": _HfExamplesInfo("Zyphra/Zamba2-7B-instruct",
                                         min_transformers_version="4.49"),
```

I think we can merge this PR first but without updating transformers (expecting users to install a later version). By setting `min_transformers_,version`, we can skip testing this model in CI for now.

Comment (yury-tokpanov) [reply tests/models/registry.py]: Sounds good

File: requirements/common.txt [hunk 6]
```diff
@@ -5,7 +5,7 @@ requests >= 2.26.0
 tqdm
 blake3
 py-cpuinfo
-transformers >= 4.48.2  # Required for Bamba model and Transformers backend.
+transformers >= 4.49.0  # Required for Zamba2 models and Transformers backend.
```
Comment (DarkLight1337) [inline requirements/common.txt:8]: Then we can revert these changes to the requirements.

Comment (yury-tokpanov) [reply requirements/common.txt]: Done

Comment (DarkLight1337) [reply requirements/common.txt]: The requirements file still shows transformers 4.49

Comment (yury-tokpanov) [reply requirements/common.txt]: ah, there test requirements, that also need to be updated - done as well

**Review by DarkLight1337 — APPROVED** [review]

**Comment by yury-tokpanov** [comment]
> Thanks, can you also update the list of supported models in the docs with this model?

Yep, all done.

Btw, I see a bunch of tests failed, but upon further inspection it seems they're unrelated to the PR.
